### DISCUSSION TITLE
Cause for non-reproducable issues with locations?

### DISCUSSION THREAD

**Discussion by synedraIT** [netbox-community/netbox#8987] [Q&A] [answered]
Hello!

This is less of a bug report and more of a question what could have gone wrong here.
We cannot include steps to reproduce this, since we are not sure ourselves what caused this. However, we are still looking into it and hopefully we can find and properly report what exactly the issue is.

Backstory: We somehow managed to introduce integrity issues (constraint violations) into our database. We could not find anything in the changelogs of the objects that we suspect caused this, but have since decided to remove them completely (the changelogs just stated that some parents and names changed around).
Now, we are facing some weird issues with locations, we are still trying to get some reproducable bug reports from this (the issues occured using just the web interface).

Here is something that is somehow possible:
 - Assign a location as parent to another location
 - You check the bottom location, it does list the parent correctly
 - The parent does NOT list the sub-location, however
 - Filtering for devices inside of a search form (/dcim/locations) WILL list the location

This is reproducable with one object, but not with another object (actually, we only found one object this IS possible with). It seems to depend on the child item being assigned, not the parent item. We already flushed our redis cache, deleted all pyc objects, fsck'd all disks and rebooted the entire host Netbox is hosted on.

As far as I understood the database, you only need to update the parent_id of the object in question, which is done correctly (we verified using the API and SQL).

Here is an excerpt of a location that should be a child of location 4:
```
{
    "id": 26,
    "url": "https://[redactednetboxurl]/api/dcim/locations/26/",
    "display": "Serverraum",
    "name": "Serverraum",
    "slug": "serverraum",
    "site": {
        "id": 1,
        "url": "https://[redactednetboxurl]/api/dcim/sites/1/",
        "display": "synedra HQ Innsbruck",
        "name": "synedra HQ Innsbruck",
        "slug": "synedra-hq-innsbruck"
    },
    "parent": {
        "id": 4,
        "url": "https://[redactednetboxurl]/api/dcim/locations/4/",
        "display": "1. OG",
        "name": "1. OG",
        "slug": "1-og",
        "_depth": 0
    },
    "tenant": null,
    "description": "",
    "tags": [],
    "custom_fields": {
        "switch_location_id": null
    },
    "created": "2022-03-28",
    "last_updated": "2022-03-29T05:24:11.682079Z",
    "rack_count": 4,
    "device_count": 45,
    "_depth": 1
}
```
However, in the webinterface, if you check the location 4, it is not listed as a sub-location:
![superlocation](https://user-images.githubusercontent.com/66822927/160546969-7d3b60ed-2fa3-442b-9b18-55c2da578b72.PNG)
Filtering the locations based on parents is working as expected and shows the correct location:
![locationlistfilter](https://user-images.githubusercontent.com/66822927/160546944-b8668d16-8463-483f-9041-5e40284aba59.PNG)
And looking at the location also shows and links the correct parent:
![sublocation](https://user-images.githubusercontent.com/66822927/160547044-49fba231-0fb2-4088-b99f-90e3350e4381.PNG)

Note: We are also experiencing Issue 8915, but we have also not found any concrete steps to reproduce this reliably yet.

We are also experiencing lots of other intermittent issues relating to parent/child locations and even location assignments, but nothing we can actually pinpoint quite as well as this.

Can anyone more experienced with netbox help us to find out what exactly is causing this or what our next steps should be in figuring this out?

**Comment by jeremystretch** [comment, marked as answer]
It sounds like the [MPTT tree](https://django-mptt.readthedocs.io/en/latest/index.html) may have gotten corrupted somehow. You can try rebuilding it using the Python shell. (This shouldn't carry much risk, but take a backup of your database before doing this just to be safe.) Activate your Python virtual environment, then execute `manage.py nbshell` and enter the following:

```
Location.objects.rebuild()
```

This will attempt to rebuild the MPTT tree from scratch for all locations.

**Reply by synedraIT** [reply]
did that and it works perfectly - thank you so much!

Just out of interest - what could have caused this? Could this also be the source of the integrity issue? Can/Should we do this for every model just to make sure, or is this some one-off event (bitflip)?

**Reply by jeremystretch** [reply]
> what could have caused this?

MPTT works by recording "left" and "right" values to maintain a hierarchy, which need to be recalculated whenever an object is added to or removed from the hierarchy. (It's pretty neat if you have the time to dig into it.) If these values become outdated because an object was added or removed outside the normal workflow, the hierarchy becomes corrupt. The good news is that it can be rebuilt using the parent assignments (as you just saw).

It's hard to say what could have caused this. It's certainly possible there's a bug in NetBox, but without known steps to reproduce it would be very difficult to track down. Most often, we see issues like this resulting from the direct manipulation of the database.

> Can/Should we do this for every model just to make sure, or is this some one-off event (bitflip)?

Only a few models in NetBox employ MPTT (locations, regions, site groups, contact groups, tenant groups, and wireless LAN groups). You can certainly re-run `rebuild()` for each of these, though it's probably not necessary.

**Reply by synedraIT** [reply]
Right, thanks for explaining. That was most likely caused by someone trying to play around with the DB (What does _this_ field do?) to reproduce/fix a bug we saw earlier, and it's unlikely to be a Bug in Netbox then.

Thanks a lot for the quick and great support!
